### DISCUSSION TITLE
🌟 React Native Maps Roadmap Discussion 🌟

### DISCUSSION THREAD

**Discussion by jan-kozinski** [react-native-maps/react-native-maps#4975] [Feature requests]
## Roadmap Discussion
Hello, React Native Maps Community! I think we can all agree that the library needs a clear roadmap and a direction to follow. 

Please help us shape it's future by sharing the features you're missing and any pain points you've encountered. Your input matters the most!

To begin, we've gone through the open discussions and collected al the requests that seemed relevant. The list, for now unordered and without timeframes goes like this:

- Add ability to load local tiles - https://github.com/react-native-maps/react-native-maps/discussions/4629
- implement onPoiClick for apple maps - https://github.com/react-native-maps/react-native-maps/discussions/4729
- Bring back web support - https://github.com/react-native-maps/react-native-maps/discussions/4741
- Vector tiles - https://github.com/react-native-maps/react-native-maps/discussions/4912
- New Architecture support - https://github.com/react-native-maps/react-native-maps/discussions/4301
- Lazy load for Callout - https://github.com/react-native-maps/react-native-maps/discussions/4409
- Huawei support - https://github.com/react-native-maps/react-native-maps/discussions/4294
- Native gradient polyline support - https://github.com/react-native-maps/react-native-maps/discussions/4744
- UI elements positioning - https://github.com/react-native-maps/react-native-maps/discussions/4733
- Polygons onPress provide coordinate - https://github.com/react-native-maps/react-native-maps/discussions/4667
- Rewrite examples in the project from class to functional components
- implement maxIntensity on Heatmap ref https://github.com/react-native-maps/react-native-maps/discussions/4649
- Add ElevationStyle Prop https://github.com/react-native-maps/react-native-maps/discussions/4633
- Add onPressStart, onPressEnd https://github.com/react-native-maps/react-native-maps/discussions/4683
- Add onPanDragStart, onPanDragEnd  https://github.com/react-native-maps/react-native-maps/discussions/4385
- create "Common Problem" section of the documentation https://github.com/react-native-maps/react-native-maps/discussions/4647
- populate testID to markers on Android https://github.com/react-native-maps/react-native-maps/discussions/4684
- add fillRule to polygon https://github.com/react-native-maps/react-native-maps/discussions/4613
- add curved lines support https://github.com/react-native-maps/react-native-maps/discussions/4596
- add ignoreTouches property to marker https://github.com/react-native-maps/react-native-maps/discussions/4550
- add AdvancedMarkers support https://github.com/react-native-maps/react-native-maps/discussions/4481
- \<MarkersCluster /\> component in pure TS for more handy markers clustering

**Comment by MariuzM** [comment]
My only gripe is when using Google on iPhone 15 Pro the maps movement is choppy / not smooth. When i tried MapBox and their map all smooth. Also Any apps that use Flutter + Google maps also smooth. Can this be improved?

**Comment by bidah** [comment]
Web is needed as more universal apps get created with expo router web support. I did play with universal apps last year with maps a bit by myself by adding the .web extension and matching API that I needed with a react.js maps counterpart, but this should be built in IMO.

**Comment by oliverloops** [comment]
I used this library a couple of times in past projects, So I'll be happy to help 👋

**Reply by jan-kozinski** [reply]
any contributions are welcome :)

**Comment by mcoppola** [comment]
@jan-kozinski Thank you for all the work maintaining this project and providing space for community input!  My company and team use this library very heavily. 

Of the priorities suggested already, these would be most impactful for us:
- New Architecture support - https://github.com/react-native-maps/react-native-maps/discussions/4301
- Add ElevationStyle Prop https://github.com/react-native-maps/react-native-maps/discussions/4633
- Add AdvancedMarkers support https://github.com/react-native-maps/react-native-maps/discussions/4481

Additionally, support for Google's emerging 3D tiles would be top of our list. 
https://developers.google.com/maps/documentation/tile/3d-tiles

One again, appreciate the engagement. I will try to advocate for some company time to contribute here.  Thanks again.

**Comment by salah-ghanim** [comment]
@jan-kozinski thanks for creating this discussion, yes this library needs a lot of effort to move it forward and it's heavily used by many companies, hopefully we'll manage to move it forward slowly and gain momentum here.

regarding the list of issues, what you've provided covers a lot of ground and many can be done gradually, what I can add:

1. Finalize the move for Animations to be done via reanimated, since there is no map specific logic there and the RN Animation API is only performant / useNativeDriver for the key properties (width, height ....)

2. introduce MapBox or OpenStreetMaps (maybe both) support to the library, Google is pushing to make maps profitable and this can limit this library usage in the future more than what's limited already.

3. Performance bottleneck analysis and improvements, some of the native code is done without having performance in mind.

4. Getting rid of the commands concept and the typing around it, there should be a simpler way to invoke native code / functions especially if/when we move to the new arch.

5. Keep the examples up to date with the changes / features we introduce.

**Comment by scriptprojectsdev** [comment]
Hi,

One thing I would add is that in iOS the pin that's used [`MKPinAnnotationView`](https://developer.apple.com/documentation/mapkit/mkpinannotationview) has been deprecated since iOS 16. This class will be removed in the future so we should move to their new default class which is [`MKMarkerAnnotationView`](https://developer.apple.com/documentation/mapkit/mkmarkerannotationview)

Thanks,

**Reply by mateki0** [reply]
Hi, @scriptprojectsdev thanks for good idea. I've opened PR with this switch - https://github.com/react-native-maps/react-native-maps/pull/5005

**Comment by salah-ghanim** [comment]
@jan-kozinski we need to add:
https://developers.google.com/maps/documentation/android-sdk/advanced-markers/overview

to the list, so that finally android markers behave like iOS (ie you can use React Views) inside the Marker to customize them.

**Comment by daniel-xyz** [comment]
Great list, I especially look forward to AdvancedMarkers. One little feature I miss for Apple Maps, which is already available on Google Maps, is to filter POI categories. https://developer.apple.com/documentation/mapkit/pointofinterestcategories

Maybe that would be good first issue for myself sometime. :D

**Comment by salah-ghanim** [comment]
more features to implement
* [ ] iOS Apple Maps: pointofInterest categories: https://developer.apple.com/documentation/mapkit/pointofinterestcategories, requested by @daniel-xyz 
* [ ] iOS GoogleMaps: add support for themes (part of latest google maps SDK)

**Comment by Stecken** [comment]
Hello @salah-ghanim! As you said to me, I'm here to discuss a new feature, I'm going to copy and paste my text here:

Currently, the Marker component allows customizing the icon via the image property, but there’s no built-in way to control the size (width and height) of the image directly. This can be inconvenient in cases where a developer wants to dynamically adjust the marker's size without having to preprocess the image dimensions in an external program.

One of the main motivations for this request came from my own experience when using the library. When I set the image property on a Marker, I expected to have more control over the image, such as being able to adjust its size dynamically. However, I realized that the only thing I could change was the URI itself (essentially the source of the image), and nothing more. This limitation felt restrictive, especially when I wanted to use a single image but scale it for different markers dynamically.

I propose adding support for width and height as optional parameters within the image property. If these parameters are provided, they would override the default dimensions of the image. If not, the image would behave as it does today, using its intrinsic size. This approach ensures backward compatibility while providing more flexibility for developers who need greater control over their custom markers.

I believe this feature would be particularly helpful for developers looking to create dynamic or customizable maps without the need to preprocess images manually. It would streamline workflows and improve the overall experience of using the library.

What do you think about this idea? I’d love to hear your feedback and thoughts before proceeding with an implementation.

**Comment by DevHamzaShahid** [comment]
Hello Team,

First of all, thank you for maintaining this amazing library! I have a feature request (or guidance request) for enhancing the navigation experience in React Native Maps.

I would like to implement the following features for a better user experience, similar to Google Maps:

Voice Assistant during Navigation

Ability to give turn-by-turn voice instructions while navigating.

Auto Camera Rotation while Navigation

Camera should rotate smoothly based on the user's heading while following a route.

Current Location Indicator with Directional Arrow & Torch Cone

Instead of a static dot, show a moving arrow with a "cone of vision" to represent the user's heading (like Google Maps).

It should rotate accurately with the compass for better orientation.

I tried using animateCamera and related methods, but the results are not as smooth or accurate as I’d like.

Questions:

Are these features already supported by this library?

If not, what would be the recommended approach or best practices for implementing them?

Are there any plans to officially add these features?

Any guidance or examples would be highly appreciated!

Thank you for your time and for all the work you’ve done on this library.
